channel-bindings enhancements

iboukris opened the pull request:
Following the discussion on krbdev (see also https://github.com/krb5/krb5/pull/1047).

Add channel-bound flag to indicate to acceptor that the bindings were checked.
Implement KERB_AP_OPTIONS_CBT from MS-KILE, to enable the initiator to signal awareness of channel-bindings.

lhoward commented on lib/gssapi/krb5/8003.c:
```diff
@@ -163,16 +163,85 @@ _gsskrb5_create_8003_checksum (
+    krb5_storage_free(sp);
+
+    return ret;
+}
```
This is very elegant but you could just use `_gss_mg_decode_le_uint32()` :)

iboukris replied:
Thanks :)

iboukris replied:
That worked.

lhoward commented on lib/krb5/build_auth.c:
```diff
@@ -73,10 +75,111 @@ make_etypelist(krb5_context context,
+
+static krb5_error_code
+store_32_le(const uint32_t in, krb5_data *out)
+{
```
This can be replaced with `_gss_mg_encode_le_uint32()`

iboukris replied:
That didn't work because libkrb doesn't link against that lib. Temporarily I copied the implementation, not sure how to go about it.

lhoward replied:
Copying the implementation is fine.

lhoward replied:
You can just inline (as in, copy and paste) the code. Obviously it shouldn't have a `_gss` prefix in libkrb5.

iboukris replied:
Thanks!

lhoward commented:
Looks like you need to export `krb5_get_ad()` on Windows too

iboukris commented:
I have a crash in client-aware on travis, which I can't reproduce locally, reviewing the code..

krb5: initiator only bindings (client-aware)
Segmentation fault (core dumped)

lhoward commented on lib/krb5/build_auth.c:
```diff
@@ -73,10 +74,93 @@ make_etypelist(krb5_context context,
+    krb5_error_code ret;
+    AuthorizationDataElement ade;
+    krb5_boolean require_cb;
+    uint32_t ap_options;
```
this should be `uint32_t ap_options[4]` if you are encoding into it

lhoward commented on lib/krb5/build_auth.c:
```diff
@@ -73,10 +74,93 @@ make_etypelist(krb5_context context,
+
+    ade.ad_type = KRB5_AUTHDATA_AP_OPTIONS;
+    ade.ad_data.length = sizeof(ap_options);
+    ade.ad_data.data = &ap_options;
```
Once you change to `uint8_t[]` be sure to remove the `&`.

iboukris commented:
Finally, the crash was not related to my changes but was due to the include logic i had in krb5.conf. So travis should pass now, while appveyor just hates me.

iboukris commented:
I settled with an alternative krb5.conf for this test, the appveyor failure looks unrelated.

lhoward commented:
krb5.conf misconfiguration likely shouldn't cause things to crash though – perhaps file a separate issue?

lhoward commented on lib/krb5/build_auth.c:
```diff
@@ -73,10 +74,93 @@ make_etypelist(krb5_context context,
+
+/* HACK: temp */
+static void
+encode_le_uint32(uint32_t n, uint8_t *p)
```
by inline in my earlier comment, I meant just put this code directly in `add_ad_options()` – it's a bit ugly but at least it's a common design pattern. Or use the `krb5_storage` API as you did originally.

lhoward commented on lib/gssapi/krb5/8003.c:
```diff
@@ -163,16 +163,64 @@ _gsskrb5_create_8003_checksum (
 }
 
+/* Flag in KRB5_AUTHDATA_AP_OPTIONS */
+#define KERB_AP_OPTIONS_CBT 0x00004000
```
might this belong in a header somewhere?

lhoward commented on lib/gssapi/gssapi/gssapi.h:
```diff
@@ -178,6 +178,8 @@ typedef OM_uint32 gss_qop_t;
 #define GSS_C_EXTENDED_ERROR_FLAG 16384
 #define GSS_C_DELEG_POLICY_FLAG 32768
 
+#define GSS_C_CHANNEL_BOUND_FLAG 2048
```
Please put this before `GSS_C_DCE_STYLE`

lhoward commented:
Can we add support for `GSS_C_MA_CBINDING_CONFIRM` and `GSS_C_MA_CBINDING_MAY_CONFIRM`, along with `GSS_C_CB_CONFIRM_FLAG`? This will require some changes to SPNEGO, I can probably help you with that (you could add `GSS_C_CB_CONFIRM_FLAG` and the attributes to `flag_to_ma_map` but the behaviour has to be change from ordering to filtering).

lhoward commented on lib/krb5/build_auth.c:
```diff
@@ -73,10 +74,93 @@ make_etypelist(krb5_context context,
+    krb5_boolean require_cb;
+    uint8_t ap_options[4];
+
+    require_cb = krb5_config_get_bool_default(context, NULL, FALSE,
```
I'm not sure whether a global configuration directive is the right way to request this. It should likely at least be per application service name (first part of SPN), and/or the initiator setting `GSS_C_CB_CONFIRM_FLAG`. @nicowilliams, you are the expert on this, what say you?

iboukris replied:
Yeah, we need a way to set it in the API, but on one hand it's not clear how the API would look like, and on the other hand we also need global config so that old applications could still use it. So I opted to start with the global config which is simpler.

See the (long) discussion on krbdev (we also discussed it on the krbdev call with Nico):
http://mailman.mit.edu/pipermail/krbdev/2020-March/013241.html

In short, by having it in global krb5.conf we can protect any existing app which operates not over TLS (say the server requires signing when not over TLS, like Windows), as it prevent a mitm to take their token and use it over TLS without signing and without channel-bindings.  Since with this flag in the token, the server will not allow all zeros when over TLS (that is, when acceptor passed bindings).

iboukris commented:
> krb5.conf misconfiguration likely shouldn't cause things to crash though – perhaps file a separate issue?

Yeah, I wasn't able to reproduce locally, only in travis CI, I'll submit a PR to demonstrate it.

iboukris commented:
> Can we add support for GSS_C_MA_CBINDING_CONFIRM and GSS_C_MA_CBINDING_MAY_CONFIRM, along with GSS_C_CB_CONFIRM_FLAG? This will require some changes to SPNEGO, I can probably help you with that (you could add GSS_C_CB_CONFIRM_FLAG and the attributes to flag_to_ma_map but the behaviour has to be change from ordering to filtering).

I'll into it.

iboukris commented:
> Can we add support for GSS_C_MA_CBINDING_CONFIRM

Looking at the draft again, this would be a much larger change, involving on the wire changes and more API which I'm not clear about. While I did take GSS_C_CHANNEL_BOUND_FLAG from there, it seemed it can be an independent feature and I was hoping we can start with having it (along with the related changes in MIT to match heimdal semantics).

On the other hand MSFT's  KERB_AP_OPTIONS_CBT ad-flag isn't compatible with the cbinding signaling proposal since it provides no confirmation to the client (if the server doesn't pass binding, the flag is ignored), so I don't see how we can merge them and we'd probably need a different  flag, like CBINDING_AWARE. But here again, I was hoping we can land the krb5.conf option first.

On a broader context, my efforts are aimed to implement support for MSFT's LdapEnforceChannelBinding flag in samba, from:
https://support.microsoft.com/en-us/help/4034879/how-to-add-the-ldapenforcechannelbinding-registry-entry

By implementing KERB_AP_OPTIONS_CBT we can support LdapEnforceChannelBinding=1, and with the addition of GSS_C_CHANNEL_BOUND_FLAG we can implement LdapEnforceChannelBinding=2 and require the bindings on the server side for all clients.

iboukris commented:
I addressed the minor comments and added man entry for client_aware_channel_bindings. Windows clients from a given version set this flag blindly on all ap-req, since they can declare that all their applications would provide bindings when over secure-channel. In open systems it is harder to claim that, but the krb5.conf option allows the admin to declare it and gain the protection.

lhoward commented:
Sorry yes, I didn't mean add support to the Kerberos mechanism for channel binding confirmation, just put the necessary mechanism attribute infrastructure in so other mechanisms could use it. But, we can do this as a separate change.

In any case, @nicowilliams really needs to review this as he understands channel bindings better than I.

iboukris commented:
Hi @nicowilliams can you take a look here?  Thanks!

nicowilliams commented on lib/gssapi/krb5/8003.c:
```diff
@@ -163,16 +163,61 @@ _gsskrb5_create_8003_checksum (
+    if (ad_len != sizeof(uint32_t))
+	return KRB5KRB_AP_ERR_MSG_TYPE;
+
+    _gss_mg_decode_le_uint32(ad_data, &ad_ap_options);
```
I guess we're going to want to be able to support more than 32 flags eventually?  Is this supposed to be a DER encoding?

nicowilliams replied:
Well, no, MIT isn't putting in the other flags.  Missed opportunity.  Maybe.  I guess we'll add any ap-options that _need_ integrity protection when we add any.

nicowilliams commented on lib/gssapi/krb5/8003.c:
```diff
@@ -163,16 +163,61 @@ _gsskrb5_create_8003_checksum (
+
+    _gss_mg_decode_le_uint32(ad_data, &ad_ap_options);
+
+    if (ad_ap_options & KERB_AP_OPTIONS_CBT)
```
Should we take this time to also use this to provide integrity protection to all other ap-options?

nicowilliams replied:
Well, no, MIT isn't putting in the other flags. Missed opportunity. Maybe. I guess we'll add any ap-options that need integrity protection when we add any.

nicowilliams commented on lib/gssapi/krb5/8003.c:
```diff
@@ -163,16 +163,61 @@ _gsskrb5_create_8003_checksum (
+	return 0;
+
+    ret = _krb5_get_ad(context, ad, NULL, KRB5_AUTHDATA_AP_OPTIONS,  &data);
+    if (ret)
```
`_krb5_get_ad()` returns `ENOENT` if it didn't find the `ad`.  Should you check for `ENOENT` here and return `0` in that case?

iboukris replied:
Alright, let me fix that.

iboukris replied:
Fixed, not sure what happened to travis but the tests pass locally.

nicowilliams commented on lib/krb5/build_auth.c:
```diff
@@ -73,10 +74,88 @@ make_etypelist(krb5_context context,
+	return ret;
+
+    /*
+     * Windows has a bug and only looks for first occurrence of AD-IF-RELEVANT
```
This is a very useful comment!  Thank you!

iboukris replied:
Here is their feedback btw:
https://lists.samba.org/archive/cifs-protocol/2020-May/003440.html

nicowilliams commented on tests/gss/new_clients_k5.conf.in:
```diff
@@ -0,0 +1,4 @@
+include @objdirabs@/krb5.conf
```
Oh, does `include` require absolute paths?

iboukris replied:
I recall it did, the man page also mentions absolute: https://github.com/heimdal/heimdal/blob/122b1720ada15ed92d229a898acfff7c6625c5bf/lib/base/config_file.c#L423

nicowilliams commented:
I did have one comment, about `_krb5_get_ad()` returning `ENOENT` when it doesn't find any `ad` of the given type.  I'll wait a bit and merge, and we can always fix it later.

iboukris commented:
Thanks all!
